File: exercism/python/run-length-encoding/run_length_encoding.py
def decode(string):
    decoded = ""

    left, right = 0, 0
    while left < len(string):
        peek = string[right]

        if peek.isdigit():
            while right < len(string) and string[right].isdigit():
                right += 1

            count = int(string[left:right])
            letter = string[right]
        else:
            count = 1
            letter = peek


        decoded += count * letter
        right += 1
        left = right

    return decoded


def encode(string):
    encoded = ""

    if not string:
        return encoded

    left, right = 0, 0

    while left < len(string):
        while right < len(string) and string[left] == string[right]:
            right += 1

        if right-left > 1:
            encoded += str(right-left) + string[left]
        else:
            encoded += string[left]

        left = right

    return encoded

File: exercism/go/run-length-encoding/run_length_encoding.go
package encode

import (
	"bytes"
	"strconv"
	"unicode"
	"unicode/utf8"
)

// RunLengthEncode encodes a string to its RLE equivalent
func RunLengthEncode(str string) string {
	var encoded bytes.Buffer
	strLen := utf8.RuneCountInString(str)

	for i := 0; i < strLen; {
		j := i
		for j < len(str) && str[i] == str[j] {
			j++
		}

		if j-i > 1 {
			encoded.WriteString(strconv.Itoa(j - i))
		}
		encoded.WriteByte(str[i])

		i = j
	}

	return encoded.String()
}

// RunLengthDecode decodes a RLE string to its full value
func RunLengthDecode(encoded string) string {
	var decoded bytes.Buffer
	encodedLen := utf8.RuneCountInString(encoded)

	for i := 0; i < encodedLen; i++ {
		j := i

		for unicode.IsDigit(rune(encoded[j])) {
			j++
		}

		times := 1
		if j > i {
			times, _ = strconv.Atoi(encoded[i:j])
			i = j
		}

		for times > 0 {
			decoded.WriteByte(encoded[i])
			times--
		}
	}

	return decoded.String()
}
